[WGSL] Implement unary expression parsing

djg opened the pull request:
#### 0141851d63984e07f592074f86cbd70f77705185
<pre>
[WGSL] Implement unary expression parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=252154">https://bugs.webkit.org/show_bug.cgi?id=252154</a>
rdar://problem/105380466

Reviewed by Tadeu Zagallo.

Handle parsing of unary expressions involving &amp;, ~, *, - and ! according to WGSL
spec.

Canonical link: <a href="https://commits.webkit.org/260232@main">https://commits.webkit.org/260232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e011c74bc70146ada94e704da1aa8b718ba65c8b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107601 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16655 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116751 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7968 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99771 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113356 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96841 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41307 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28468 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9646 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29820 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10317 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49402 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7072 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11877 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -161,6 +161,20 @@ static bool canContinueRelationalExpression(const Token& token)
     }
 }
 
+static bool canProceedSingularExpression(const Token& token)
```
nit: should this be called `canProceedUnaryExpression` to match the rest?

djg replied:
Well a unary expression included the operator, ie `-x` is a unary expression, so I named it for the `x`. It should probably be called `canBeginUnaryExpression`.

tadeuzagallo commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -840,10 +873,11 @@ Expected<UniqueRef<AST::Expression>, Error> Parser<Lexer>::parseUnaryExpression(
-    if (current().m_type == TokenType::Minus) {
+    if (canProceedSingularExpression(current())) {
+        auto op = toUnaryOperation(current());
         consume();
```
Random thought, but maybe consume should also return `Token&`? That way you could write `toUnaryOperation(consume())`

djg replied:
It would have to return a `Token` right? `consume()` bumps the internal storage of what the current token is.

tadeuzagallo reviewed (approved):
r=me

djg committed:
[WGSL] Implement unary expression parsing
https://bugs.webkit.org/show_bug.cgi?id=252154
rdar://problem/105380466

Reviewed by Tadeu Zagallo.

Handle parsing of unary expressions involving &, ~, *, - and ! according to WGSL
spec.

Canonical link: https://commits.webkit.org/260232@main
